### PR TITLE
duobert for msmarco (interactive and script tested), support for interactive trec-car,  update to recent pyserini, include temp colab link in README

### PR DESCRIPTION
**Pull request by ronakice**
Update colab/requirements. 
Include link to temp colab in README.md

### CODE REVIEW

**Comment by ronakice** [comment]
The same colab link as earlier has the duoBERT example too.

File: numbert/reranker.py [hunk 1]
```diff
@@ -8,6 +8,12 @@
 from .utils.utils_numbert import (processors, output_modes,
                                     convert_examples_to_features)
 
+try:
+    from tokenizers import BertWordPieceTokenizer
```
Comment (rodrigonogueira4) [inline numbert/reranker.py]: Maybe print a warning in case `tokenizers` is not available? Also, maybe just add `tokenizers` to the requirements and let python raise an exception in case it is not installed?

Comment (ronakice) [reply numbert/reranker.py]: I basically wanted it to run without the need for tokenizers (mostly because few places like ComputeCanada seem to have an issue getting tokenizers running so it should default to the regular slower tokenizer in that case. But since tokenizers is already in the requirements, I should probably just remove this.

Comment (ronakice) [reply numbert/reranker.py]: Fixed in latest commit.

File: numbert/utils/utils_numbert.py [hunk 2]
```diff
@@ -229,100 +231,147 @@ def merge(self, qrels, run, queries):
...
+    def get_train_examples(self, data_dir, is_qrels = None, is_duoBERT = False):
         """See base class."""
 
         return self._create_examples_train_triples(self.load_train_triples(
                 os.path.join(data_dir,"triples.train.small.tsv")),
-                "train")
+                "train", is_duoBERT)
 
-    def get_train_examples_old(self, data_dir, is_qrels = True):
+    def get_train_examples_old(self, data_dir, is_qrels = True, is_duoBERT = False):
```
Comment (rodrigonogueira4) [inline numbert/utils/utils_numbert.py]: What is this used for? Maybe rename it `_v1`or `deprecated`?

File: numbert/utils/utils_numbert.py [hunk 3]
```diff
@@ -229,100 +231,147 @@ def merge(self, qrels, run, queries):
...
+                for doc_ind_b, doc_b in enumerate(data[qid]):
+                    docid_dict[doc_b.docid] = convert_to_unicode(doc_b.content)
+                    for doc_ind_c, doc_c in enumerate(data[qid]):
+                        if doc_ind_b == doc_ind_c:
+                            continue
+                        guid = "%s-%s-%s-%s-%s-%s" % ("online", qid, doc_ind_b, doc_b.docid, doc_ind_c, doc_c.docid)
+                        text_b = convert_to_unicode(doc_b.content)
+                        text_c = convert_to_unicode(doc_c.content)
+                        examples.append(
+                            InputExample(guid=guid, text_a=text_a, text_b=text_b, text_c=text_c, label=str(0)))
```
Comment (rodrigonogueira4) [inline numbert/utils/utils_numbert.py]: Maybe add a line break here so it within the 80 char-limit? Same for the ones below.

File: numbert/utils/utils_numbert.py [hunk 4]
```diff
@@ -430,7 +479,7 @@ def get_examples_online(self, queries, data):
                 text_b = convert_to_unicode(doc.content)
                 docid_dict[doc.docid] = text_b
                 examples.append(
-                    InputExample(guid=guid, text_a=text_a, text_b=text_b, label=str(0)))
+                    InputExample(guid=guid, text_a=text_a, text_b=text_b, label=str(0), len_gt_titles=42))
```
Comment (rodrigonogueira4) [inline numbert/utils/utils_numbert.py]: Why is it 42? Maybe add a comment explaining this?

File: numbert/utils/utils_numbert.py [hunk 5]
```diff
@@ -526,35 +579,84 @@ def convert_examples_to_features(examples, tokenizer,
...
+            tokens_b = batch_tokens_b[ex_index%batch_size]
+        elif is_tokenizers:
+            tokens_a = tokenizer.encode(example.text_a).ids
+            tokens_b = tokenizer.encode(example.text_b).ids[1:]
+        else:
+            tokens_a = tokenizer.tokenize(example.text_a)
+            tokens_b = tokenizer.tokenize(example.text_b)
+        tokens = tokens_a[:max_len_A - int(sep_token_extra) + int(is_tokenizers)] 
+        if is_tokenizers:
+            tokens[-1] = 102
```
Comment (rodrigonogueira4) [inline numbert/utils/utils_numbert.py:652]: Maybe add a comment explaining what is `102`?

**Review by rodrigonogueira4 — COMMENTED** [review]
Awesome! Looks great overall. Just a couple of nits

File: numbert/utils_reranker.py [hunk 6]
```diff
@@ -87,15 +95,24 @@ def online_eval(model, tokenizer, device, features, batch_size = 1, model_type="
...
     for ind, guid in enumerate(lguids):
-        # if out_label_ids[ind] == 1:
-        #     if guid[1] in numbert_labels:
-        #         numbert_labels[guid[1]].add(guid[3])
-        #     else:
-        #         numbert_labels[guid[1]] = {guid[3]}
-        if guid[1] in numbert_predictions:
-            numbert_predictions[guid[1]].append((guid[3], preds[ind][1]))
+        if is_duoBERT:
+            if guid[1] in numbert_predictions:
```
Comment (rodrigonogueira4) [inline numbert/utils_reranker.py]: This block is pretty hard to parse. Maybe rewrite it or add comments explaining what it does?

File: numbert/utils/utils_numbert.py (hunk 2 above)
Comment (ronakice) [reply numbert/utils/utils_numbert.py]: This is for using the other train file provided by MSMARCO instead of the triples one. I'll rename it as  *_non_triples. Right now this isn't being used but potentially can use this in the future.

**Comment by ronakice** [comment]
@rodrigonogueira4 I think I've gone over all your points and addressed them in the files in the framework. Thanks for reviewing in detail! Let me know if anything else is wrong.

**Review by rodrigonogueira4 — APPROVED** [review]
Looks all good to me. Thanks for implementing this!
